### PR TITLE
add mixtool list functionality

### PR DESCRIPTION
**Pull request by allenmqcymp**
@metalmatze could you take a look at this?

Tabwriter isn't fully working because the lines don't wrap if the description is too long.
Otherwise I think it works as a PoC.

### CODE REVIEW

File: cmd/mixtool/list.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	"net/url"
+	"os"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/pkg/errors"
+
+	"github.com/urfave/cli"
```
Comment (metalmatze) [inline cmd/mixtool/list.go:31]: Remove the empty line and run goimports again :)

File: cmd/mixtool/list.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+			cli.StringFlag{
+				Name:  "path, p",
+				Usage: "provide the path of a url with a json endpoint or a local json file",
+			},
+		},
+	}
+}
+
+func queryWebsite(mixinsWebsite string) ([]byte, error) {
+
```
Comment (metalmatze) [inline cmd/mixtool/list.go:61]: No need for a new line after the func declaration.

File: cmd/mixtool/list.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	client := http.Client{
+		Timeout: time.Second * 3,
```
Comment (metalmatze) [inline cmd/mixtool/list.go:63]: 10 seconds should be fine too. 3s might be a bit little in some networks.

File: cmd/mixtool/list.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+	req.Header.Set("User-Agent", "mixtool-list")
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
```
Comment (metalmatze) [inline cmd/mixtool/list.go:79]: Interesting. I don't think you need to wrap the call here. The docs in the standard lib don't have this either:
```golang
resp, err := http.Get("http://example.com/")
if err != nil {
	// handle error
}
defer resp.Body.Close()
body, err := ioutil.ReadAll(resp.Body)
```

File: cmd/mixtool/list.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	return body, nil
+}
+
+func listAction(c *cli.Context) error {
+
+	// if path is not specified, default to the mixins website
+	// otherwise, try parse as url
+	// otherwise, try look for a local json file
+
+	defaultWebsite := "https://monitoring.mixins.dev/mixins.json"
```
Comment (metalmatze) [inline cmd/mixtool/list.go:95]: I think this is actually a perfect example for a `const defaultWebsite` at the very top of the file. :)

File: cmd/mixtool/list.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+	path := c.String("path")
+	var body []byte
+	var err error
+	if path == "" {
+		body, err = queryWebsite(defaultWebsite)
+		if err != nil {
+			return err
+		}
```
Comment (metalmatze) [inline cmd/mixtool/list.go:104]: I'd do it the other way around. Using early returns.
```suggestion
	if path != "" {
	...
		return nil
	}
```

That way there's no need for the else clause.

File: cmd/mixtool/list.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+			}
+		} else {
+			body, err = queryWebsite(path)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	var mixins map[string][]mixin
```
Comment (metalmatze) [inline cmd/mixtool/list.go:126]: You have created this exact same type at the top of the file, why not use it here too?
In the end, I prefer simply using the map inline. I don't think there's a need for the `type mixin { map[string][]mixin }` as you demonstrate yourself here. :)

File: cmd/mixtool/list.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	jsonErr := json.Unmarshal(body, &mixins)
+
+	if jsonErr != nil {
+		return errors.New("failed to unmarshal json. Maybe your json schema is incorrect?")
+	}
```
Comment (metalmatze) [inline cmd/mixtool/list.go:131]:
```suggestion
	if err := json.Unmarshal(body, &mixins); err != nil {
		return fmt.Errorf("failed to unmarshal json: %w", err)
	}
```

**Review by metalmatze — CHANGES_REQUESTED** [review]
Left a couple of comments on your code.

The descriptions are an interesting problem. Maybe it's enough for now to only show the name of the mixin? Otherwise, because we control the JSON endpoint, we can think about writing a summary next to the description that's just one sentence, so it should fit most screens.

File: cmd/mixtool/list.go (hunk 7 above)
Comment (allenmqcymp) [reply cmd/mixtool/list.go:126]: Yeah I think the inline type is better too! I think that was an oversight. :/

File: cmd/mixtool/list.go (hunk 8 above)
Comment (allenmqcymp) [reply cmd/mixtool/list.go:131]: OK. Out of interest, what's the guide for when to use errors.New vs fmt.Errorf?

File: cmd/mixtool/list.go (hunk 6 above)
Comment (allenmqcymp) [reply cmd/mixtool/list.go:104]: Here's the logic I think. Now from here I'm not quite sure how to simplify the control flow. Not sure how to get rid of the else clauses entirely, that is.

```
If path == "":
  err = queryWebsite(default)
  if err:
    return err
else: (need to check if its a url or localfile)
  if path is a valid url:
    err = queryWebsite(url)
    if err:
      return err
  else:
    if path a valid localfile:
      err = readlocalfile()
      if err:
        return err

```

**Comment by allenmqcymp** [comment]
> The descriptions are an interesting problem. Maybe it's enough for now to only show the name of the mixin? Otherwise, because we control the JSON endpoint, we can think about writing a summary next to the description that's just one sentence, so it should fit most screens.

Sure, I left out the description for now (commented it out). Lucas and I agreed that we can work on polishing the UI later, since tabwriter might not be the best tool if we really needed to show extended description. I also like the short description alternative though.

**Comment by allenmqcymp** [comment]
@metalmatze ready for a second pass :)

File: cmd/mixtool/list.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+// if path is not specified, default to the mixins website
+// otherwise, try parse as url
+// otherwise, try look for a local json file
+func listAction(c *cli.Context) error {
+	path := c.String("path")
+	var body []byte
+	var err error
+	if path == "" {
+		body, err = queryWebsite(defaultWebsite)
+		if err != nil {
```
Comment (metalmatze) [inline cmd/mixtool/list.go:88]:
```suggestion
		body, err = queryWebsite(defaultWebsite)
		if err != nil {
```
```suggestion
		body, err = queryWebsite(defaultWebsite)
		return err
```

File: .gitignore [hunk 10]
```diff
@@ -1,2 +1,3 @@
 /_output/
 /mixtool
+.vscode
```
Comment (metalmatze) [inline .gitignore:3]: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: cmd/mixtool/list.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	if err != nil {
+		return err
+	}
+
+	mixins, err := parseMixinJSON(body)
+	if err != nil {
+		return err
+	}
+	return printMixins(mixins)
+
```
Comment (metalmatze) [inline cmd/mixtool/list.go:153]: No need for this empty line.

File: cmd/mixtool/server.go [hunk 12]
```diff
@@ -84,6 +84,7 @@ func (h *ruleProvisioningHandler) ServeHTTP(w http.ResponseWriter, r *http.Reque
 	}
 
 	if reloadNecessary {
+		fmt.Println("reloading prometheus")
```
Comment (metalmatze) [inline cmd/mixtool/server.go:87]: Don't want debug printlns in the final commit. Please remove :)

**Review by metalmatze — APPROVED** [review]
Very nice!

**Pull request merged**